lib/dataset/test/coords_view_test.cpp

```diff
@@ -124,6 +127,113 @@ TEST(MutableCoordsViewTest, item_write) {
   ASSERT_EQ(coords[Dim::Y], y_reference);
 }
 
+TEST(SizedDictTest, default_init) {
+  const Coords coords;
+  ASSERT_TRUE(coords.empty());
+  ASSERT_EQ(coords.sizes(), Sizes());
+  ASSERT_FALSE(coords.sizes_are_set());
```

Comment:
Leaving aside how one would implement this on the C++ side, I would feel much better if `Dataset().coords` would return `None` if there is no data. This whole `sizes_are_set` mechanism looks like it was no fun at all to implement, risks some odd bugs and is hard to think about. What is your view on this?

Reply:
Would you still return this even when coords have been assigned? E.g.
```python
ds = sc.Dataset(coords={'x': sc.scalar(1)})
```

Reply:
No, then they are not None.

Reply:
Then this would no longer work:
```python
ds = sc.Dataset(coords={'x': sc.scalar(1)})
ds['a'] = sc.arange('x', 6)
```
Here, inserting a scalar coord does not set the sizes. So we can insert an array data item. To support this, we need a way to track whether the dataset is supposed to contain scalars or whether the coord does not relate to the data's dims.

Reply:
I think that is fine? The sizes of the coords is explicitly `{}`, I do not see a real problem?

Reply:
Yes, I think so?

Reply:
It is conceptually a bit strange because normally, scalar coords don't relate to the size of the dataset / data array. But I don't think this is a strong reason for supporting it here.

One last argument against your suggestion: It would make type checking annoying. Every use of `ds.coords[key]` would be flagged as a bad union access because `None` doesn't support the operation. There is no easy way around it for the user. They would have to either write lengthy code to check the type, cast to `Mapping[str, sc.Variable]`, or suppress the check in this line.

Reply:
Type checking is a good argument. I suppose we have the same problem for, among others, `da.unit` and `da.bins`. Not sure what to do about all these?

Reply:
Hmm, but can't you apply the same argument to uninitialized `sizes` (unless you represent them as empty dict)?

Reply:
Discussion summary:
- Revert addition of `sizes` arg to `Dataset.__init__`
- Require either data or coords passed to `Dataset.__init__`, such that we can always define sizes.
  - `Dataset({})` -> `sizes={}`
  - `Dataset()` -> exception
  - `Dataset(coords={})` -> `sizes={}`
- If unknown sizes, use `dict` or `DataGroup` and construct `Dataset` after adding items.